### ISSUE TITLE
Change of behavior in 1.2 makes passing tests with 1.1 fail with 1.2

### ISSUE THREAD

**Issue by b-per** [dbt-labs/dbt-utils#923]
### Describe the bug

The fix of the issue https://github.com/dbt-labs/dbt-utils/issues/785 in the PR #765 introduces a change of behavior when the 2 tables of the comparison do not contain the same columns.

This led to the fact that integration tests in `dbt_project_evaluator` passed with `dbt_utils` 1.1 but failed with `1.2`

I wouldn't have expected a minor bump to change the behavior.

**Comment by dbeatty10** [comment]
I see what you are saying in regards to behavior change @b-per:
- In 1.1, there was **not** an error when the column names didn't match
- In 1.2, there **is** an error when the column names don't match

It looks like [this](https://github.com/dbt-labs/dbt-utils/blob/9e2b498028c03c13056e4ea58ece1368d4857458/CHANGELOG.md?plain=1#L18) is the root cause:

> - Equality test will now raise an error when the second model has less columns than the first by @rlh1994 in [#765](https://github.com/dbt-labs/dbt-utils/pull/765)


Out of curiosity, what does it do in 1.1 if you move the equality test from the seed to the model?

i.e., move [this](https://github.com/dbt-labs/dbt-project-evaluator/blob/6ef7a69214b3b8ce3e503f9dd88b20d4ffae6598/integration_tests/seeds/tests/tests_seeds.yml#L6-L11) test to [here](https://github.com/dbt-labs/dbt-project-evaluator/blob/6ef7a69214b3b8ce3e503f9dd88b20d4ffae6598/models/marts/tests/testing.yml#L20) instead (making sure to remove the `exclude_columns` part)? Does it error in 1.1? Or does it still succeed?

**Comment by b-per** [comment]
In 1.1 changing the side of the comparison fails with this error:

```
Binder Error: Referenced column "resource_type" not found in FROM clause!
```

as we were expecting.

I am not saying that we should revert #785 but I guess this is more of a discussion whether `dbt_utils.equality` working when the main table has less columns than the compared one was the intended behavior or an unintended and untested one.

**Comment by dbeatty10** [comment]
### TLDR

The original behavior looks like an oversight to me rather than an intentional decision that fully considered all the implications.

### Context

The behavior of `dbt_utils.equality` working when the main table has less columns than the compared one has existed since the 2nd commit in dbt utils: https://github.com/dbt-labs/dbt-utils/commit/948dd9b8ee2abf966ca46a48005329eec332a58c

I don't know if it was intended or unintended, but I buy the argument in https://github.com/dbt-labs/dbt-utils/issues/785 that _by default_ `dbt_utils.equality` should have always held the following properties:
- [commutativity](https://en.wikipedia.org/wiki/Commutative_property)
- same definition as equality in [set theory](https://en.wikipedia.org/wiki/Set_(mathematics)#Naive_set_theory):
  - A=B if and only if:
    - A is a subset of B
    - B is a subset of A
  
### Behavior change

This behavior did change in 1.2.0, but we're considering this a backward-compatible bug fix: an internal change that fixes [incorrect behavior](https://semver.org/#spec-item-6). Since it didn't introduce any backward incompatible API changes, this doesn't require bumping the major version.

dbt 1.7+ has [predictable package installs](https://docs.getdbt.com/reference/commands/deps#predictable-package-installs) by default via `package-lock.yml` which should help mitigate any pre-existing projects breaking overnight in a surprising way.

### Summary

Upgrading to dbt-utils 1.2.0 may require code changes for some projects, but we're considering this a good thing because it fixes a bug.
